File: lab2/src/T1.java
import java.util.Scanner;
import java.util.concurrent.BrokenBarrierException;

public class T1 extends Thread {
    private final Resources resources;

    private final int begin;
    private final int end;

    public T1(Resources resources) {
        setName("T1");
        this.resources = resources;

        begin = 0;
        end = begin + resources.h;
    }

    @Override
    public void run() {
        System.out.println("Thread " + getName() + " started");
        try {
            Thread.sleep(200);
            Main.CS1.lock(); // input x and C
            Scanner in = new Scanner(System.in);
            System.out.printf("%s. Enter value of x: ", getName());
            resources.x = in.nextInt();

            System.out.printf("%s. Enter value to fill vector C: ", getName());
            resources.C = Data.fillVectorValue(resources.n, in.nextInt());
            Main.CS1.unlock();
            Main.B1.await(); // wait for other threads to input their data

            // calculate ah
            int a1 = Data.vectorMultiply(resources.B, resources.C, begin, end);
            resources.a.addAndGet(a1);

            // get shared Matrix MB
            Main.S1.acquire();
            int[][] MB1 = resources.MB;
            Main.S1.release();

            // calculate MTh
            int[][] MT1 = Data.matrixMultiply(resources.MA, MB1, begin ,end);
            Data.makeMatrix(resources.MT, MT1, begin, end);

            // Signal to threads T2, T3, T4 about calculating MTh
            Main.S2.release(3);
            // Waiting for signal from T2
            Main.S3.acquire();
            // Waiting for signal from T3
            Main.S4.acquire();
            // Waiting for signal from T4
            Main.S5.acquire();

            // get shared x
            Main.CS2.lock();
            int xh = resources.x;
            Main.CS2.unlock();

            // get shared a
            Main.S6.acquire();
            int a = resources.a.get();
            Main.S6.release();

            // calculate Zh: Dh * a + (E * MTh) * x
            int[] Zh = Data.addVectors(Data.vectorScalarMultiply(resources.D, a, begin, end),
                                        Data.vectorScalarMultiply(Data.vectorMatrixMultiply(resources.E, resources.MT, begin, end), xh));
            Data.makeFinalVector(resources.Z, Zh, begin, end);

            // Waiting for T2, T3, T4 to calculate Zh
            Main.B2.await();

            // Output the result
            System.out.print(getName() + ". Vector Z: ");
            Data.printVector(resources.Z);
        } catch (InterruptedException | BrokenBarrierException ex) {
            System.out.printf("%s is interrupted.\n", getName());
            Thread.currentThread().interrupt();
        }
    }
}

File: lab3/T1.cs
using System.Diagnostics;

namespace lab3 {
    public class T1 {
        private Resources resources;

        private int begin;
        private int end;
        
        public T1(Resources resources) {
            this.resources = resources;

            begin = 0;
            end = begin + resources.h;
        }

        public void Run() {
            Console.WriteLine("Thread T1 started");
            try {
                // input data
                Data.FillMatrix(resources.MB, 1);
                Data.FillMatrix(resources.MC, 1);

                // Wait for T4 to input data
                Program.S1.Release(3);
                Program.S2.WaitOne();

                // calculation 1: ai = min(Zh)
                int ai = Data.FindMinimum(resources.Z, begin, end);
                // calculation 2 and CS1: a = min(a, ai)
                Interlocked.Exchange(ref resources.a, Math.Min(resources.a, ai));

                // Signal to threads T2, T3, T4 about calculating a
                Program.E1.Set();
                // Waiting for signal from T2
                Program.E2.WaitOne();
                // Waiting for signal from T3
                Program.E3.WaitOne();
                // Waiting for signal from T4
                Program.E4.WaitOne();

                // CS2: get shared d
                int d;
                lock (Program.Lock1) {
                    d = resources.d;
                }

                // CS3: get shared a
                Program.M1.WaitOne();
                int a = resources.a;
                Program.M1.ReleaseMutex();
                
                // calculation3 : MOh = MB * (MCh * MM) * d + a * MCh
                int[,] MOh = Data.MatrixAddition(Data.MatrixScalarMultiply(Data.MatrixMultiply(Data.MatrixMultiply(resources.MC, resources.MM, begin, end), resources.MB), d), Data.MatrixScalarMultiply(resources.MC, a, begin, end));
                Data.MakeFinalMatrix(resources.MO, MOh, begin, end);

                // Waiting for T2, T3, T4 to calculate MOh
                Program.B1.SignalAndWait();
            } catch (ThreadInterruptedException) {
                Console.WriteLine("Thread T1 interrupted");
                Thread.CurrentThread.Interrupt();
            }
        }
    }
}